solidity/contracts/bridge/Frauds.sol

```diff
@@ -0,0 +1,577 @@
+    ///         proves a given transaction input unlocking the given UTXO was
+    ///         actually signed by the wallet. This function cannot determine
+    ///         whether the transaction was actually broadcast and the input was
+    ///         consumed in a fraudulent way so it just opens a challenge period
+    ///         during which the wallet can defeat the challenge by submitting
+    ///         proof of a transaction that consumes the given input according
+    ///         to protocol rules. To prevent spurious allegations, the caller
+    ///         must deposit ETH that is returned back upon justified fraud
+    ///         challenge or confiscated otherwise
+    /// @param walletPublicKey The public key of the wallet in the uncompressed
+    ///        and unprefixed format (64 bytes)
+    /// @param sighash The hash that was used to produce the ECDSA signature
+    ///        that is the subject of the fraud claim. This hash is constructed
+    ///        by applying double SHA-256 over a serialized subset of the
+    ///        transaction. The exact subset used as hash preimage depends on
+    ///        the transaction input the signature is produced for. See BIP-143
+    ///        for reference
+    /// @param signature Bitcoin signature in the R/S/V format
+    function submitFraudChallenge(
+        Data storage self,
+        bytes memory walletPublicKey,
+        bytes32 sighash,
+        RSVSignature calldata signature
+    ) external {
+        require(
+            msg.value >= self.challengeDepositAmount,
+            "The amount of ETH deposited is too low"
+        );
+
+        require(
+            CheckBitcoinSigs.checkSig(
+                walletPublicKey,
+                sighash,
+                signature.v,
+                signature.r,
+                signature.s
+            ),
+            "Signature verification failure"
+        );
+
+        uint256 challengeKey = uint256(
+            keccak256(
+                abi.encodePacked(
+                    walletPublicKey,
+                    sighash,
+                    signature.v,
+                    signature.r,
+                    signature.s
+                )
+            )
+        );
+
+        FraudChallenge storage challenge = self.challenges[challengeKey];
+        require(challenge.reportedAt == 0, "Fraud challenge already exists");
+
+        challenge.challenger = msg.sender;
+        challenge.depositAmount = msg.value;
+        /* solhint-disable-next-line not-rely-on-time */
+        challenge.reportedAt = uint32(block.timestamp);
+        challenge.closed = false;
+
+        bytes memory compressedWalletPublicKey = EcdsaLib.compressPublicKey(
+            walletPublicKey.slice32(0),
+            walletPublicKey.slice32(32)
+        );
+        bytes20 walletPubKeyHash = bytes20(compressedWalletPublicKey.hash160());
+
+        emit FraudChallengeSubmitted(
+            walletPubKeyHash,
```

Comment:
`walletPubKeyHash` is actually the main key used to identify wallets in the `Bridge`. I think we should rather limit the usage of wallet ID and use it only in places where we interact with the ECDSA registry. We already go this way within `Wallets` library.

Reply:
Yeah, that's a good point. `DepositRevealed` and `RedemptionRequested` also uses `walletPubKeyHash`. 

So maybe we can just pass the wallet pub key hash as a parameter to the library so that it does not have to be recomputed again?

Reply:
Done in c9739b02241816f217c05ae941b30bf52ce40c0b.
The public key is passed and not calculated again.